src/map.cpp

```diff
@@ -532,62 +532,150 @@ bool Map::canThrowObjectTo(const Position& fromPos, const Position& toPos, bool
 	return isSightClear(fromPos, toPos, false);
 }
 
-bool Map::checkSightLine(const Position& fromPos, const Position& toPos) const
+//get a line between two points
+//based on Xiaolin Wu algorithm
+std::vector< std::pair<int, int> > get2DLine(int x0, int y0, int x1, int y1)
 {
-	if (fromPos == toPos) {
-		return true;
+	std::vector< std::pair<int, int> > line;
+	if (x0 == x1 && y0 == y1) {
+		return line;
 	}
 
-	Position start(fromPos.z > toPos.z ? toPos : fromPos);
-	Position destination(fromPos.z > toPos.z ? fromPos : toPos);
+	bool isSteep = std::abs(y1 - y0) > std::abs(x1 - x0);
 
-	const int8_t mx = start.x < destination.x ? 1 : start.x == destination.x ? 0 : -1;
-	const int8_t my = start.y < destination.y ? 1 : start.y == destination.y ? 0 : -1;
+	if (isSteep) {
+		std::swap(x0, y0);
+		std::swap(x1, y1);
+	}
 
-	int32_t A = Position::getOffsetY(destination, start);
-	int32_t B = Position::getOffsetX(start, destination);
-	int32_t C = -(A * destination.x + B * destination.y);
+	if (x0 > x1) {
+		std::swap(x0, x1);
+		std::swap(y0, y1);
+	}
 
-	while (start.x != destination.x || start.y != destination.y) {
-		int32_t move_hor = std::abs(A * (start.x + mx) + B * (start.y) + C);
-		int32_t move_ver = std::abs(A * (start.x) + B * (start.y + my) + C);
-		int32_t move_cross = std::abs(A * (start.x + mx) + B * (start.y + my) + C);
+	float dx = x1 - x0;
+	float dy = y1 - y0;
+	float grad = 1.0;
 
-		if (start.y != destination.y && (start.x == destination.x || move_hor > move_ver || move_hor > move_cross)) {
-			start.y += my;
-		}
+	if (dx != 0) {
+		grad = dy * 1.0 / dx;
+		// We need to offset the line for isometric layout
+		grad = grad + .03;
```

Comment:
Where does `0.03` come from?

Reply:
Probably from testing the code and finding the right value, but it's just my wild guess.

Reply:
> Probably from testing the code and finding the right value, but it's just my wild guess.

Yes, it was after a lot of researching the different grad values that were being used in each spot. What I gathered was the issue was taking place if the grad was: 0.3333 repeating. If the grad was anything else it was fine. That value is just enough that it won't interfere with the way the line is reading in any other scenario. It is hard to explain without you seeing all grad values I printed out after trying to throw a coin in every spot open.

It's definitely not optimal. A better way would be to add the offset only in the specific instance its needed. This is what I was able to come up with as an alternative. Not that its detrimental to anything.

Reply:
It would be good for someone to check it with otclient and see if the line remains the same past the limits of the tibia client but that's not what TFS is for.

Reply:
I think the ideal solution would be to mirror the vector to always be in the same direction, then calculate only one variation i.e. normalize every distance to the positive vector (top-left to bortom-right) and check that without magic numbers.

Reply:
I recommend testing this on both Linux and Windows anyway because they handle calculations a bit differently (for example math.random gets different results on same seed).